spline_background.py

```diff
@@ -32,57 +35,45 @@ def bspline_background(variable, dim, smoothing_factor=None):
     errors = variable.variances
     weights = 1.0/errors
 
-    # https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.UnivariateSpline.html
+    # flag marking 'bin-edge' status of input x values
+    bin_edge_input = False
+
+    # If input x values are bin-edges, use bin centres to spline
+
+    if len(x_values) == len(values)+1:
+        bin_edge_input = True
+
+        new_x_values = [0.5*(x_values[i] + x_values[i+1])
+                        for i in range(len(x_values)-1)]
+        x_values = new_x_values
+
     # find out the knots and splines.
-    knots, u = interpolate.splprep([x_values, values], s=smoothing_factor, k=5, w=weights)
+    knots, u = interpolate.splprep([x_values, values],
+                                   s=smoothing_factor,
+                                   k=5,
+                                   w=weights)
     # perform the B-spline interpolation
     splined = interpolate.splev(u, knots)
     # splined is a [2][data_length] array for X and Y values
 
     # cast splined into DataArray type
-    output_x = sc.Variable(dims=[dim], values=splined[0])
+    if bin_edge_input:
+        # redefine output x_values to be 'bin-edges'
+        len_splined = len(splined[0])
+        bin_edge_output_x = np.zeros(len_splined + 1)
+        bin_edge_output_x[0] = 0.5*(splined[0][0] - splined[0][1])
+        bin_edge_output_x[len_splined] = 0.5*(
+                                         3*splined[0][len_splined-1] -
+                                         splined[0][len_splined-2])
+
+        for i in range(len_splined-1):
+            bin_edge_output_x[i+1] = 0.5*(splined[0][i+1] - splined[0][i])
+
+        output_x = sc.Variable(dims=[dim], values=bin_edge_output_x)
+    else:
+        output_x = sc.Variable(dims=[dim], values=splined[0])
+
```

Comment:
```suggestion
    output_x = variable.coords[dim]
```

Reply:
The input x-values cannot be simply copied to the output because the scipy algorithms modify it.

Reply:
Ah, ok. You can still shorten by doing parts of the math with numpy arrays though.